cirq-google/cirq_google/named_topologies.py

```diff
@@ -0,0 +1,295 @@
+        to NetworkX plotting functionality.
+    """
+    if ax is None:
+        ax = plt.gca()  # coverage: ignore
+
+    if cartesian:
+        pos = {node: (y, -x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+    else:
+        pos = {node: (x + y, y - x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+
+    nx.draw_networkx(graph, pos=pos, ax=ax, **kwargs)
+    ax.axis('equal')
+    return pos
+
+
+@dataclass(frozen=True)
+class LineTopology(NamedTopology):
+    """A 1D linear topology.
+
+    Node indices are contiguous integers starting from 0 with edges between
+    adjacent integers.
+
+    Args:
+        n_nodes: The number of nodes in a line.
+    """
+
+    n_nodes: int
+
+    def __post_init__(self):
+        if self.n_nodes <= 1:
+            raise ValueError("`n_nodes` must be greater than 1.")
+        object.__setattr__(self, 'name', f'line-{self.n_nodes}')
+        graph = nx.from_edgelist(
+            [(i1, i2) for i1, i2 in zip(range(self.n_nodes), range(1, self.n_nodes))]
+        )
+        object.__setattr__(self, 'graph', graph)
+
+    def draw(self, ax=None, cartesian: bool = True, **kwargs) -> Dict[Any, Tuple[int, int]]:
+        """Draw this graph.
+
+        Args:
+            ax: Optional matplotlib axis to use for drawing.
+            cartesian: If True, draw as a horizontal line. Otherwise, draw on a diagonal.
+            kwargs: Additional arguments to pass to `nx.draw_networkx`.
+        """
+        g2 = nx.relabel_nodes(self.graph, {n: (n, 1) for n in self.graph.nodes})
+        return draw_gridlike(g2, ax=ax, cartesian=cartesian, **kwargs)
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+
+@dataclass(frozen=True)
+class DiagonalRectangleTopology(NamedTopology):
+    """A grid topology forming a rectangle rotated 45-degrees.
+
+    This topology is based on Google devices where plaquettes consist of four qubits in a square
+    connected to a central qubit:
+
+        x   x
+          x
+        x   x
+
+    The corner nodes are not connected to each other. `width` and `height` refer to the number
+    of unit cells, or equivalently the number of central nodes. Each unit cell contributes
+    two nodes when in bulk: the central node and 1/4 of each of the four shared corner nodes.
+    Accounting for the boundary, the total number of nodes is 2*w*h + w + h + 1. An example
+    diagram showing the diagonal nature of the rectangle is reproduced below. It is a
+    "diagonal-rectangle-3-2" with width 3 and height 2. This can be most clearly seen by focusing
+    on the central nodes diagrammed with an `x`, of which there are 2x3=6. The `*` nodes are
+    added to ensure each `x` node has degree four.
+
+          *
+         *x*
+        *x*x*
+         *x*x*
+          *x*
+           *
+
+    In the surface code, the `*` nodes are data qubits and the `x` nodes are measure qubits.
```

Comment:
The roles can in fact be swapped, so I'd either drop this comment or say that qubits with a given marker have the same role (data vs measure) in the surface code.

Reply:
removed